File: ch3/queue_via_stacks/java/Queue.java
import java.util.*;

public class Queue<Obj>{
    private Stack<Obj> q;
    private int size;
    Queue(){
        q = new Stack<Obj>();
        size = 0;
    }
    public void enqueue(Obj n){
        Stack<Obj> tmp = new Stack<Obj>();
        for(int i = 0; i < size; i++){
            tmp.push(q.peek());
            q.pop();
        }
        q.push(n);
        for(int i = 0; i < size; i++){
            q.push(tmp.peek());
            tmp.pop();
        }
        size++;
    }
    public Obj dequeue(){
        Obj tmp = q.peek();
        q.pop();
        size--;
        return tmp;
    }
}

File: ch3/queue_via_stacks/python/Queue.py
class Queue:
    def __init__(self):
        self.queue = []

    def enqueue(self, n):
        tmp = []
        for i in range(len(self.queue)):
            tmp.append(self.queue.pop())
        self.queue.append(n)
        for i in range(len(tmp)):
            self.queue.append(tmp.pop())

    def dequeue(self):
        return self.queue.pop()

    def p(self):
        for i in range(len(self.queue)):
            print(self.queue[i], end=" ")
        print("")

def main():
    q = Queue()
    q.enqueue(1)
    q.enqueue(2)
    q.enqueue(3)
    q.enqueue(4)
    q.enqueue(5)
    q.enqueue(6)
    q.enqueue(7)

    q.p()

    q.dequeue()

    q.p()

if __name__ == "__main__":
    main()
